src/qx/Simulator.cpp

```diff
@@ -2,78 +2,105 @@
 #include <iostream>
 #include <vector>
 
 
 namespace qx {
 
-std::optional<SimulationError> Simulator::set(std::string const &filePath) {
+using V1AnalysisResult = cqasm::v1x::analyzer::AnalysisResult;
+using V1Program = cqasm::v1x::ast::One<cqasm::v1x::semantic::Program>;
+
+using V3AnalysisResult = cqasm::v3x::analyzer::AnalysisResult;
+using V3Program = cqasm::v3x::ast::One<cqasm::v3x::semantic::Program>;
+
+namespace {
+V1AnalysisResult parseCqasmV1xFile(std::string const &filePath) {
+    auto analyzer = cqasm::v1x::default_analyzer("1.2");
+
+    return analyzer.analyze_file(filePath);
+}
+
+V1AnalysisResult parseCqasmV1xString(std::string const &s) {
     auto analyzer = cqasm::v1x::default_analyzer("1.2");
 
-    auto analysisResult = analyzer.analyze(filePath);
+    return analyzer.analyze_string(s);
+}
+
+V3AnalysisResult parseCqasmV3xFile(std::string const &filePath) {
+    auto analyzer = cqasm::v3x::default_analyzer("3.0");
+
+    return analyzer.analyze_file(filePath);
+}
+
+V3AnalysisResult parseCqasmV3xString(std::string const &s) {
+    auto analyzer = cqasm::v3x::default_analyzer("3.0");
+
+    return analyzer.analyze_string(s);
+}
+
+std::variant<V1Program, SimulationError> getV1ProgramOrError(V1AnalysisResult const& analysisResult) {
     if (!analysisResult.errors.empty()) {
-        program.reset();
         std::stringstream error;
-        error << "Cannot parse and analyze file " << filePath << ":";
+        error << "Cannot parse and analyze cQASM v1: \n";
 
-        std::for_each(analysisResult.errors.begin(), analysisResult.errors.end(), [&error](auto const& s) { error << "\n" << s; });
+        std::for_each(analysisResult.errors.begin(), analysisResult.errors.end(), [&error](auto const& x) { error << x << "\n"; });
 
         return SimulationError{error.str()};
     }
 
-    program = analysisResult.root;
+    auto program = analysisResult.root;
     assert(!program.empty());
 
     if (!program->error_model.empty() &&
         program->error_model->name != "depolarizing_channel") {
-        program.reset();
+        std::cerr << "Unknown error model: " << program->error_model->name
+                << std::endl;
         return SimulationError{"Unknown error model: " + program->error_model->name};
     }
 
-    return std::nullopt;
+    return program;
 }
 
-std::optional<SimulationError> Simulator::setString(std::string const &s) {
-    auto analyzer = cqasm::v1x::default_analyzer("1.2");
-
-    auto analysisResult = analyzer.analyze_string(s);
+std::variant<V3Program, SimulationError> getV3ProgramOrError(V3AnalysisResult const& analysisResult) {
     if (!analysisResult.errors.empty()) {
-        program.reset();
         std::stringstream error;
-        error << "Cannot parse and analyze string " << s << ": \n";
+        error << "Cannot parse and analyze cQASM v3: \n";
 
         std::for_each(analysisResult.errors.begin(), analysisResult.errors.end(), [&error](auto const& x) { error << x << "\n"; });
```

Comment:
There's a beautiful C++ idiom for doing something like this ([demo](https://godbolt.org/z/xT46d1rsa)):

```
std::copy(analysisResult.errors.begin(), analysisResult.errors.end(), std::ostream_iterator{ std::cout, "\n" });
```

But it's not suitable here because of the `error <<`.

Reply:
Done with a `fmt::format` plus `fmt::join`.